File: src/main/java/me/fru1t/streamtools/javafx/SettingsController.java
package me.fru1t.streamtools.javafx;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import javafx.fxml.FXML;
import javafx.scene.control.Button;
import javafx.stage.Stage;
import me.fru1t.javafx.Controller;
import me.fru1t.streamtools.Settings;

import java.lang.reflect.ParameterizedType;
import java.util.HashSet;

/**
 * A specialized controller that controls a settings panel that requires a save and cancel button.
 * @param <T> The object type passed back to event listeners.
 */
public abstract class SettingsController<T extends Settings<T>> extends Controller {
    /**
     * Possible events for the SettingsController.
     */
    public interface EventHandler<T extends Settings<T>> {
        void onSettingsChange(T settings);
    }

    // FXML Declarations
    private @FXML Button settingsCancelButton;
    private @FXML Button settingsSaveButton;
    private @FXML Button settingsApplyButton;

    // Class declarations
    protected final T currentSettings;
    private final HashSet<EventHandler<T>> eventHandlers;

    protected SettingsController() {
        // Setup event handlers
        eventHandlers = new HashSet<>();

        // Set up currentSettings
        // Because we aren't handed any other context to the settings, we must infer the Settings
        // class through the generic T, then ask for its @DefaultSetting annotation to pull default
        // settings field for that class.

        // Grab this class (SettingsController<T>) as a parametrized type from the implementing
        // class's definition.
        ParameterizedType settingsControllerClass =
                (ParameterizedType) getClass().getGenericSuperclass();

        // Get the class type T
        @SuppressWarnings("unchecked")
        Class<T> settingsClass = (Class<T>) settingsControllerClass.getActualTypeArguments()[0];

        try {
            Constructor<T> settingsConstructor = settingsClass.getConstructor();
            currentSettings = settingsConstructor.newInstance();
        } catch (NoSuchMethodException | IllegalAccessException | InstantiationException | InvocationTargetException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * When overriding this method, this parent method must be called using
     * <code>super.onSceneCreate();</code>
     */
    @Override
    protected void onSceneCreate() {
        super.onSceneCreate();

        // Check that the settings panel has both a cancel and save button action.
        if (settingsCancelButton == null) {
            throw new RuntimeException(getFxmlResourcePath() + " requires a button with an fx:id "
                    + "of \"settingsCancelButton\"");
        }
        if (settingsSaveButton == null) {
            throw new RuntimeException(getFxmlResourcePath() + " requires a button with an fx:id "
                    + "of \"settingsSaveButton\"");
        }
        if (settingsApplyButton == null) {
            throw new RuntimeException(getFxmlResourcePath() + " requires a button with an fx:id "
                    + "of \"settingsApplyButton\"");
        }

        // Wire cancel and save buttons
        settingsCancelButton.setOnAction(event -> SettingsController.this.getStage().hide());
        settingsApplyButton.setOnAction(event -> {
            commitSettings();
            for (EventHandler<T> handler : eventHandlers) {
                handler.onSettingsChange(currentSettings);
            }
        });
        settingsSaveButton.setOnAction(event -> {
            commitSettings();
            for (EventHandler<T> handler : eventHandlers) {
                handler.onSettingsChange(currentSettings);
            }
            SettingsController.this.getStage().hide();
        });

        // All controls should lose focus when the root is clicked
        getScene().getRoot().setOnMouseClicked(event -> getScene().getRoot().requestFocus());
    }

    @Override
    public void show() {
        refresh();
        super.show();
    }

    @Override
    public void onStageProvide(Stage stage) {
        super.onStageProvide(stage);
        stage.setResizable(false);
    }

    // Adds an event handler
    public void addEventHandler(EventHandler<T> handler) {
        eventHandlers.add(handler);
    }

    // Removes an event handler
    public void removeEventHandler(EventHandler<T> handler) {
        eventHandlers.remove(handler);
    }

    /**
     * @return The currently committed settings.
     */
    public T getCurrentSettings() {
        return currentSettings.copy();
    }

    /**
     * Updates the settings to the given settings object.
     * @param settings The new settings
     */
    public void update(Settings<?> settings) {
        currentSettings.update(settings);
        refresh();
    }

    /**
     * Fired when the save button is clicked. The implementor should commit all GUI changes to
     * the {@link #currentSettings} object.
     */
    protected abstract void commitSettings();

    /**
     * Updates the GUI to reflect the currentSettings object.
     */
    protected abstract void refresh();
}

File: src/main/kotlin/me/fru1t/streamtools/javafx/SettingsController.kt
package me.fru1t.streamtools.javafx

import java.lang.reflect.InvocationTargetException
import javafx.fxml.FXML
import javafx.scene.control.Button
import javax.annotation.OverridingMethodsMustInvokeSuper
import me.fru1t.javafx.Controller
import me.fru1t.kotlin.ConcurrentWeakReferenceHashSet
import me.fru1t.streamtools.controller.Settings
import java.lang.ref.WeakReference

import java.lang.reflect.ParameterizedType

/**
 * A controller that manages a [Settings] object specified by [T] which is updated through the GUI.
 * Layouts that extend this base class must include a Cancel, Save, and Apply button.
 */
abstract class SettingsController<T : Settings<T>> protected constructor() : Controller() {

  // FXML Declarations
  @FXML private lateinit var settingsCancelButton: Button
  @FXML private lateinit var settingsSaveButton: Button
  @FXML private lateinit var settingsApplyButton: Button

  // Class declarations
  val currentSettings: T

  private val onSettingsChangeListeners: ConcurrentWeakReferenceHashSet<(T) -> Unit> =
      ConcurrentWeakReferenceHashSet()

  init {
    // We need our own copy of settings, in order to do that, we just create a new instance of
    // T (our settings type).

    // Grab this class (SettingsController<T>) as a parametrized type from the implementing
    // class's definition.
    val settingsControllerClass = javaClass.genericSuperclass as ParameterizedType

    // Get the class type T
    @Suppress("UNCHECKED_CAST")
    val settingsClass = settingsControllerClass.actualTypeArguments[0] as Class<T>

    // Create an instance of it with the no-arg constructor
    try {
      val settingsConstructor = settingsClass.getConstructor()
      currentSettings = settingsConstructor.newInstance()
    } catch (e: NoSuchMethodException ) {
      throw RuntimeException(e)
    } catch (e: IllegalAccessException) {
      throw RuntimeException(e)
    } catch (e: InstantiationException) {
      throw RuntimeException(e)
    } catch (e: InvocationTargetException) {
      throw RuntimeException(e)
    }

    // Add ourselves to the list of event handlers
    addOnSettingsChangeListener(::onSettingsChange)
  }

  @OverridingMethodsMustInvokeSuper
  override fun onSceneCreate() {
    super.onSceneCreate()

    // Kotlin's lateinit allows us to denote fields as non-null, but they may be null. We want to
    // fail early if this happens.
    @Suppress("SENSELESS_COMPARISON")
    if (settingsCancelButton == null) {
      throw RuntimeException(
          "${getFxmlResourcePath()} requires a button with an fx:id of \"settingsCancelButton\"")
    }
    @Suppress("SENSELESS_COMPARISON")
    if (settingsSaveButton == null) {
      throw RuntimeException(
          "${getFxmlResourcePath()} requires a button with an fx:id of \"settingsSaveButton\"")
    }
    @Suppress("SENSELESS_COMPARISON")
    if (settingsApplyButton == null) {
      throw RuntimeException(
          "${getFxmlResourcePath()} requires a button with an fx:id of \"settingsApplyButton\"")
    }

    // Wire cancel and save buttons
    settingsCancelButton.setOnAction { _ -> this@SettingsController.stage!!.hide() }
    settingsApplyButton.setOnAction { _ ->
      commitSettings()
      onSettingsChangeListeners.forEach { it(currentSettings.copy()) }
    }
    settingsSaveButton.setOnAction { _ ->
      commitSettings()
      onSettingsChangeListeners.forEach { it(currentSettings.copy()) }
      this@SettingsController.stage!!.hide()
    }

    // All controls should lose focus when the root is clicked
    scene.root.setOnMouseClicked { _ -> scene.root.requestFocus() }
  }

  @OverridingMethodsMustInvokeSuper
  override fun show() {
    stage?.isResizable = false
    onSettingsChange(currentSettings)
    super.show()
  }

  /**
   * Adds a listener that is fired every time these settings change. The listener is passed a copy
   * of the settings. Listeners are stored as [WeakReference]s, so they will be automatically
   * garbage collected once the underlying object is gone.
   */
  fun addOnSettingsChangeListener(listener: (T) -> Unit) {
    onSettingsChangeListeners.add(listener)
  }

  /** Updates the current settings with the given [settings], making a defensive copy. */
  fun update(settings: Settings<*>) {
    currentSettings.update(settings)
    onSettingsChangeListeners.forEach { it(currentSettings.copy()) }
  }

  /** Commit all GUI changes to the [currentSettings] object. */
  protected abstract fun commitSettings()

  /** Updates the GUI from the [currentSettings] object (or [settings]). */
  protected abstract fun onSettingsChange(settings: T)
}
